java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -0,0 +1,228 @@
+                put("name", schema.hash)
+                put("is_primitive", false)
+            }
+            val schemaTypeId = insert("types", null, content)
+
+            schemaTypeMap.value[schema.hash] = schemaTypeId
+
+            val insertFieldStatement = compileStatement(
+                "INSERT INTO fields (type_id, parent_type_id, name) VALUES (?, ?, ?)"
+            )
+            val insertFieldBlock = { fieldName: String, fieldType: FieldType ->
+                insertFieldStatement.apply {
+                    bindLong(1, getTypeId(fieldType))
+                    bindLong(2, schemaTypeId)
+                    bindString(3, fieldName)
+                    executeInsert()
+                }
+                Unit
+            }
+            schema.fields.singletons.forEach(insertFieldBlock)
+            schema.fields.collections.forEach(insertFieldBlock)
+            schemaTypeId
+        }
+    }
+
+    /** Returns the type ID for the given [fieldType] if known, otherwise throws. */
+    @VisibleForTesting
+    fun getTypeId(fieldType: FieldType): TypeId = when (fieldType) {
+        is FieldType.Primitive -> fieldType.primitiveType.ordinal.toLong()
+        is FieldType.EntityRef -> schemaTypeMap.value[fieldType.schemaHash]
+            ?: throw IllegalArgumentException(
+                "Unknown type ID for schema with hash ${fieldType.schemaHash}"
+            )
+    }
+
+    /** Loads all schema type IDs from the 'types' table into memory. */
+    private fun loadTypes(): MutableMap<String, TypeId> {
+        val schemaTypeMap = mutableMapOf<String, TypeId>()
+        readableDatabase.rawQuery(
+            "SELECT name, id FROM types WHERE is_primitive = 0",
+            emptyArray()
+        ).use {
+            while (it.moveToNext()) {
+                val hash = it.getString(0)
+                val id = it.getLong(1)
+                schemaTypeMap[hash] = id
+            }
+        }
+        return schemaTypeMap
+    }
+
+    companion object {
+        internal const val DB_NAME = "resurrection.sqlite3"
+        private const val DB_VERSION = 1
+
+        private val CREATE =
+            """
+                CREATE TABLE types (
+                    id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                    name TEXT NOT NULL UNIQUE,
+                    is_primitive INTEGER NOT NULL DEFAULT 0
+                )
+
+                CREATE INDEX type_name_index ON types (name, id)
+
+                CREATE TABLE storage_keys (
+                    id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                    storage_key TEXT UNIQUE NOT NULL
+                )
+
+                CREATE INDEX storage_key_index ON storage_keys (storage_key, id)
+
+                CREATE TABLE entities (
+                    storage_key_id INTEGER NOT NULL PRIMARY KEY,
+                    type_id INTEGER NOT NULL
+                )
+
+                CREATE TABLE collections (
```

Comment:
I think we may need singletons too?

```sql
CREATE TABLE singletons (
    storage_key_id INTEGER NOT NULL PRIMARY KEY,
    type_id INTEGER NOT NULL,
    value_id INTEGER # Allow nulls
)
```

Where `value_id` would point to the entity record if the `type_id` is non-primitive, and to a primitive value id if it _is_ a primitive.

Reply:
Thanks, done